### PR TITLE
Maptext 2023: I can see clearly now

### LINKED ISSUE #73002: Small Fonts (our maptext font) needs to be replaced
"Small Fonts" doesn't come with all versions of Windows. I don't know what it comes with, but I have seen many screenshots of people not having it and the text looking terrible.

We can't legally distribute it as it is a builtin Windows font, and the license is not apparent. Theoretically possible to get a license from Microsoft, but fat chance.

We have VCR OSD Mono now, but it doesn't look good at very small sizes, only big ones.

Whatever font we replace it with needs to be non-blurry at the *very* small sizes maptext is usually at (Runechat, balloon alerts), which is extremely hard to find a font for. It must also be public domain, or otherwise distributable with our license. I don't really have any suggestions.

### PR DESCRIPTION
**Pull request by lessthnthree**
## About The Pull Request

We needed to replace our maptext font. Closes https://github.com/tgstation/tgstation/issues/73002

Replaced with a pair of new fonts, made some improvements and fixes while we're at it.

- Two new maptext fonts: Grand9K and TinyUnicode
- These two scale cleanly with BYOND icon sizes, so no antialiasing of already tiny fonts making it look bad when enlarged on screen
- Fixed size fonts now have metrics to be used properly on overlays (such as status display) where a client (a requirement for MeasureText) is not feasible
- VCR OSD Mono still here, usable as a variable size font with larger text
- Synchronizes status display scrolling speed between lines
- Fixed special maptexts for yelling, clown, redtext, greentext

https://github.com/tgstation/tgstation/assets/83487515/7dbaf1d1-f1d5-46ff-8903-d95cab059ed8

## Why It's Good For The Game

Maptext looks like crap if you're one of the players who no longer have the old font. Even if you do have Small Fonts, this looks better.

## Changelog

:cl: LT3
refactor: Refactored maptext (those floating words)
fix: Fixed special chat bubbles for yelling, clown, redtext, greentext
fix: Fixed alignment of status display text
code: Status displays now synchronize their message lines when scrolling
/:cl:

### CODE REVIEW

File: code/__DEFINES/fonts.dm [hunk 1]
```diff
@@ -14,3 +14,11 @@
 
 /// Emoji icon set
 #define EMOJI_SET 'icons/ui_icons/emoji/emoji.dmi'
+
+// Font metrics bitfield
+/// Include leading A width and trailing C width in GetWidth() or in DrawText()
+#define INCLUDE_AC (1<<0)
```
Comment (Mothblocks) [inline code/__DEFINES/fonts.dm:20]: I don't know what this means at all

File: interface/fonts/fonts_datum.dm [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	var/ascent      	// distance above baseline (including whitespace)
+	var/descent     	// distance below baseline
+	var/avgwidth    	// average character width
+	var/maxwidth    	// maximum character width
+	var/overhang    	// extra width, such as from italics, for a line
+	var/inleading   	// internal leading vertical space, for accent marks
+	var/exleading   	// external leading vertical space, just plain blank
+	var/defchar     	// default character (for undefined chars)
+	var/start       	// first character in metrics
+	var/end				// last character in metrics
```
Comment (Mothblocks) [inline interface/fonts/fonts_datum.dm:20]: This is not how we comment these.

Make these comments before each variable with `///`.

File: interface/fonts/fonts_datum.dm [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+			firstline=0
+			idx=0
+			continue
+		idx=(ch-start)*3
+		if(idx<=0 || idx>=metrics.len) idx=(defchar-start)*3
+		if(!. && !(flags&INCLUDE_AC)) .-=metrics[idx+1]
+		.+=metrics[idx+1]+metrics[idx+2]+metrics[idx+3]
+	if(. && idx && !(flags&INCLUDE_AC)) .-=max(metrics[idx+3],0)
+	.=max(.+firstline,longest)
+	if(.>0) .+=overhang
```
Comment (Mothblocks) [inline interface/fonts/fonts_datum.dm:48]: No no no no no no no, make this at all readable

File: interface/fonts/fonts_datum.dm [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	/// The font file we link to.
+	var/font_family
+	/// Font features and metrics
+	/// Generated by Lummox's dmifontsplus (https://www.byond.com/developer/LummoxJR/DmiFontsPlus)
+	var/list/metrics
+	var/height      	// total height of a line
+	var/ascent      	// distance above baseline (including whitespace)
+	var/descent     	// distance below baseline
+	var/avgwidth    	// average character width
+	var/maxwidth    	// maximum character width
```
Comment (Mothblocks) [inline interface/fonts/fonts_datum.dm:14]: `average_width`, `max_width`

File: interface/fonts/grand_9k.dm [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+	descent = 2
+	avgwidth = 4
+	maxwidth = 9
+	overhang = 0
+	inleading = 4
+	exleading = 1
+	defchar = 31
+	start = 30
+	end = 255
+	metrics = list(\
```
Comment (Mothblocks) [inline interface/fonts/grand_9k.dm:24]: You don't need to backslash lists

File: interface/fonts/grand_9k.dm [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+		0, 5, 1,	/* char 30 */ \
```
Comment (Mothblocks) [inline interface/fonts/grand_9k.dm:25]: Never tab align, just a normal `0, 5, 1, // char 30`

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Looks great!

File: code/__DEFINES/fonts.dm (hunk 1 above)
Comment (lessthnthree) [reply code/__DEFINES/fonts.dm:20]: This is all from Lummox and I'm aware it has to be cleaned up. I wanted to get the PR up for the rest of it to be reviewed in the meantime.

**Review by Imaginos16 — APPROVED** [review]
Looks nice. I'm pleasantly surprised to see something I didn't ask for that looks very visually pleasing!

**Comment by Mothblocks** [comment]
Also can you please post the licenses to both these fonts so we know we can use them?

**Comment by lessthnthree** [comment]
> Also can you please post the licenses to both these fonts so we know we can use them?

It's in the updated license.txt with VCR OSD Mono

**Comment by Kapu1178** [comment]
How are these fonts actually served to users?

**Comment by lessthnthree** [comment]
> How are these fonts actually served to users?

In the RSC.

**Comment by Kapu1178** [comment]
> > How are these fonts actually served to users?
> 
> In the RSC.

Fonts shipped in the RSC are just usable to clients? Neat.

**Comment by lessthnthree** [comment]
> > > How are these fonts actually served to users?
> > 
> > 
> > In the RSC.
> 
> Fonts shipped in the RSC are just usable to clients? Neat.

Not drag and drop, you need to make a font datum in DM and reference the .ttf

**Comment by san7890** [comment]
> > Also can you please post the licenses to both these fonts so we know we can use them?
> 
> It's in the updated license.txt with VCR OSD Mono

The only license.txt file touched in this PR is a deletion of the file:

image

Probably unstaged file?

**Comment by LemonInTheDark** [comment]
do these work with unicode or are our russian forks fucked

**Comment by lessthnthree** [comment]
> do these work with unicode or are our russian forks fucked

I don't speak Russian, but I think with the right CSS nothing is stopping it, it has the characters.

image

**Comment by lessthnthree** [comment]
> do these work with unicode or are our russian forks fucked

Yep, included in the two fonts.

image

**Comment by LemonInTheDark** [comment]
Yoooo

**Comment by Fikou** [comment]
that looks really aliased. are you sure that's correct?

**Comment by lessthnthree** [comment]
> that looks really aliased. are you sure that's correct and not like arial or something?

It could be a fallback. Nothing stops a downstream from adding a Russian friendlier modular font. It's still an improvement.

For context, this is the current Russian maptext on master.

image

**Comment by SinfulBliss** [comment]
This looks like shit

**Comment by AnywayFarus** [comment]
Это пипец.

File: tgstation.dme [hunk 7]
```diff
@@ -1,6 +1,7 @@
 // DM Environment file for tgstation.dme.
 // All manual changes should be made outside the BEGIN_ and END_ blocks.
 // New source code should be placed in .dm files: choose File/New --> Code File.
+// interface\skin.dmf needs to be at the very end of the file.
```
Comment (LemonInTheDark) [inline tgstation.dme:4]: why? byond appears to be trying to force it to the top, we should not have people editing in dreammaker cause breakage to the dme. 

please work out a way to work around this if it is what I suspect, rather then enforcing it with a line of text

Comment (lessthnthree) [reply tgstation.dme:4]: Okay.

Comment (LemonInTheDark) [reply tgstation.dme:4]: thanks. I don't wanna be too aggressive here, this is a real cool change, I just really don't like deprecating dreammaker, which seems? like is what's going on

Comment (Mothblocks) [reply tgstation.dme:4]: didnt see this either, not good =(

**Comment by lessthnthree** [comment]
This is old, it's gone already.

**Pull request merged**